### ISSUE TITLE
Add "let" in the term type ?

### ISSUE THREAD

**Issue by fblanqui** [Deducteam/lambdapi#208]
Currently, there is a let in the LP syntax but it is translated into a beta-redex at scoping time. Hence, sharing is lost. We could improve this by having an explicit let in the term type. But keeping this sharing as much as possible is difficult when doing reduction, unification and type inference. This could be a possible future work.

**Comment by rlepigre** [comment]
There is also an issue with dependent types when `let` is implemented as  a `redex`. But the purpose of this extension was really to get more "readable" code. In any case, I'm not sure whether it is really useful to go further that that. We will have to see if such a construct is really useful in practice, given the other features of Lambdapi.

By the way @fblanqui, have you developed more proofs in the system?

**Comment by rlepigre** [comment]
As I said in #222, if the `let` is defined as a redex (and it currently is), then it is typed as follows.
```
     Γ, x:A ⊢ u : B
  -------------------- ∀i
   Γ ⊢ λx.u : ∀x∈A, B         Γ ⊢ t : A
  -------------------------------------- ∀e
         Γ ⊢ (λx.u) t : B[x := t]
     ================================
      Γ ⊢ let x = t in u : B[x := t]
```
However, this is too weak because in the premisse for `u` you lose the equation `x ≡ t`.

The expected rule for the `let` should b something like the following.
```
   Γ, x:A ⊢ u : B[x := t]       Γ ⊢ t : A
  ----------------------------------------
       Γ ⊢ let x = t in u : B[x := t]
```
Or alternatively, carrying the equality constraint `x ≡ t` in the context (I think that this is what is done in Coq).

**Comment by rlepigre** [comment]
> Currently, there is a let in the LP syntax but it is translated into a beta-redex at scoping time. Hence, sharing is lost. 

@fblanqui: this has never been a sharing problem, but a limitation in expressiveness. In other words, there is a rule that is more general and still sound.

**Comment by francoisthire** [comment]
Maybe it would have been better to keep the other issue on and closed this one since the other one was more informative.

I agree with @rlepigre it is not a problem of sharing here, or at least, I don't see how. As I said in the previous issue #222 , this as also an impact on the translators to Dedukti.

**Comment by gabrielhdt** [comment]
solved by fda8752
